### ISSUE TITLE
OrientDB 2.0 - $stack problem

### ISSUE THREAD

**Issue by johndavids** [orientechnologies/orientdb#3476]
Hello, I'm using OrientDB 2.0
I have a simple traverse query

`select $stack, $path from (traverse out() from #12:6)`

and the result is

`$stack	$path`
`[]	(#12:6)`
`[]	(#12:6).out[0](#12:5)`
`[]	(#12:6).out[0](#12:5).out[0](#12:4)`
`[]	(#12:6).out[0](#12:5).out[1](#12:2)`
`[]	(#12:6).out[0](#12:5).out[2](#13:0)`

The $path is accurate, but how come the $stack is empty?

**Comment by adam-ho** [comment]
Same on 2.1RC. 
What is the status of this guys? 
Others seem to be having the same issue: http://orient-database.narkive.com/HENVrXur/orientdb-extracting-vertices-and-edges-from-the-path-variable

**Comment by lvca** [comment]
Just tried with the OpenBeer database:
![image](https://cloud.githubusercontent.com/assets/312606/8105421/48f77ff6-1039-11e5-848a-580a20bfeedd.png)

**Comment by adam-ho** [comment]
The problem is still there, please do not close the bug.

The query that does work:
```
select @rid, $stack from (traverse * from (select from Service where Name like '%nternatio%'))
```

Queries that do not work:
```
select @rid, $stack from (traverse * from (select from Service where Name like '%nternatio%')) where @rid = #13:11

select from (select @rid, $stack from (traverse * from (select from Service where Name like '%nternatio%'))) where rid = #13:11
```

Looks like if you use a where to filter out a specific node you lose the information about the stack.
This is troublesome as I'm trying to find the path between two specific nodes and not interested in paths between any other nodes.

**Comment by adam-ho** [comment]
To reproduce on the Beer database:

The query that does work:
```
select @rid, $path, $stack from (traverse both() from #12:58) 
select $path, $stack from (traverse both() from #12:58 while $depth < 4)
```

Queries that do not work:
```
select @rid, $path, $stack from (traverse both() from #12:58) where @rid = #11:4
select @rid, $path, $stack from (traverse both() from #12:58 while $depth < 4) where @rid = #11:4
```

**Comment by lvca** [comment]
I close the bug to avoid "walking issues" around. Reopened.

**Comment by sparkycrisse** [comment]
+1 Also experiencing these $stack issues

**Comment by lvca** [comment]
I guess the reason is the auto-optimization of the query execution. This query:

```sql
select @rid, $path, $stack from (traverse both() from #12:58) where @rid = #11:4
```

Is identical to:

```sql
select @rid, $path, $stack from #11:4
```

So OrientDB execution planner skip the traverse at all. Could you write here a query that has any significance for your app and we try to figure out if the problem it's this or any others?

**Comment by adam-ho** [comment]
I've already did in the first post:
```
select @rid, $stack from (traverse * from (select from Service where Name like '%nternatio%')) where @rid = #13:11

select from (select @rid, $stack from (traverse * from (select from Service where Name like '%nternatio%'))) where rid = #13:11
```
Both of them trying to do the same thing:
Start from a node called "international" and evaluate all paths. When you find the path to the node called #13:11 (or several other), tell me how you got there.

I'm still unsure why you believe 
```
select @rid, $path, $stack from (traverse both() from #12:58) where @rid = #11:4
```
Is identical to:
```
select @rid, $path, $stack from #11:4
```
Because they aren't.
Neither logically nor structurally. Under what circumstances would it not be a bug to drop a whole subquery or an instruction (traverse()) in a query?

**Comment by luigidellaquila** [comment]
Hi @johndavids 

Fixed in develop branch, the fix will be released in v 2.2.1

Thanks

Luigi
